domain/schema/model.go

```diff
@@ -281,33 +281,395 @@ ON application (name);
+        REFERENCES      charm(uuid),
+    PRIMARY KEY (charm_uuid, name)
+);
+
+CREATE INDEX idx_charm_device_charm
+ON charm_device (charm_uuid);
+
+CREATE TABLE charm_payload (
+    charm_uuid          TEXT NOT NULL,
+    name                TEXT,
+    type                TEXT,
+    CONSTRAINT          fk_charm_payload_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    PRIMARY KEY (charm_uuid, name)
+);
+
+CREATE INDEX idx_charm_payload_charm
+ON charm_payload (charm_uuid);
+
+CREATE TABLE charm_resource_kind (
+    id       INT PRIMARY KEY,
+    name     TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_charm_resource_kind_name
+    ON charm_resource_kind (name);
+
+INSERT INTO charm_resource_kind VALUES
+    (0, 'file'),
+    (1, 'oci-image');
+
+CREATE TABLE charm_resource (
+    charm_uuid          TEXT NOT NULL,
+    name                TEXT,
+    kind_id             INT NOT NULL,
+    path                TEXT,
+    description         TEXT,
+    CONSTRAINT          fk_charm_resource_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    CONSTRAINT          fk_charm_resource_kind
+        FOREIGN KEY     (kind_id)
+        REFERENCES      charm_resource_kind(id),
+    PRIMARY KEY (charm_uuid, name)
+);
+
+CREATE INDEX idx_charm_resource_charm
+ON charm_resource (charm_uuid);
+
+CREATE TABLE charm_term (
+    charm_uuid          TEXT NOT NULL,
+    value               TEXT NOT NULL,
+    CONSTRAINT          fk_charm_term_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    PRIMARY KEY (charm_uuid, value)
+);
+
+CREATE INDEX idx_charm_term_charm
+ON charm_term (charm_uuid);
+
+CREATE TABLE charm_container (
+    charm_uuid          TEXT NOT NULL,
+    name                TEXT,
+    resource            TEXT,
+    uid                 INT,
+    gid                 INT,
+    CONSTRAINT          fk_charm_container_charm
+        FOREIGN KEY     (charm_uuid)
+        REFERENCES      charm(uuid),
+    PRIMARY KEY (charm_uuid, resource)
+);
+
+CREATE INDEX idx_charm_container_charm
+ON charm_container (charm_uuid);
+
+CREATE TABLE charm_container_mount (
+    charm_uuid            TEXT NOT NULL,
+    charm_container_name  TEXT,
+    resource              TEXT,
+    storage               TEXT,
+    location              TEXT,
+    CONSTRAINT            fk_charm_container_mount_charm
+        FOREIGN KEY       (charm_uuid)
+        REFERENCES        charm(uuid),
+    CONSTRAINT            fk_charm_container_mount_charm_container
+        FOREIGN KEY       (charm_container_name)
+        REFERENCES        charm_container(name),
+    PRIMARY KEY (charm_uuid, resource)
+);
+
+CREATE INDEX idx_charm_container_mount_charm
+ON charm_container_mount (charm_uuid);
+
+-- Create a charm url view for backwards compatibility.
+CREATE VIEW v_charm_url AS
+SELECT cs.name || ':' || c.name || '-' || IFNULL(co.revision, 0) AS url, c.uuid AS uuid
```

Comment:
I think `SELECT CONCAT(...) AS url...` is more readable, but it turns out CONCAT isn't available everywhere, so we need to fallback to the concat operator (`||`) instead.